♻️ Rewrite react-string-replace and refactor it

Benjamin-Piper opened the pull request:
This is a rewrite and refactor and has no effect on the website. Especially because this code is not integrated yet. That's coming in #98 

We discussed after submitting #97 that the API of [react-string-replace] was limiting. We decided to write our own implementation to perform the replacements of the kit functions. This let us include capture groups as replacer function parameters and leads to, IMO, cleaner code in general.

This PR also installs Preact as it's a JSX component framework that is smaller than React.

[react-string-replace]: https://github.com/iansinnott/react-string-replace

**Considered alternative: [string-replace-callback]**

This would have been nice, almost perfect, but we didn't give this a shot because [the npm package is not up to date with the GitHub release][issue].

[string-replace-callback]: https://github.com/cardboardshark/string-replace-callback
[issue]: https://github.com/cardboardshark/string-replace-callback/issues/1

Benjamin-Piper committed:
➕ Install @preact/preset-vite

Benjamin-Piper committed:
♻️ Rename regex function ignore to not_in_set

Benjamin-Piper committed:
♻️ Implement react-string-replace ourselves

Benjamin-Piper committed:
♻️ Add new regex consts for characters

Benjamin-Piper committed:
➕ Install htmr and preact-render-to-string

We realised that react-string-replace had a limiting syntax. We wanted to
be able to access the captured strings, so we're implementing our own
version of that function.

Benjamin-Piper committed:
➖ Uninstall react-string-replace

Benjamin-Piper committed:
➕ Install @types/react

Benjamin-Piper committed:
🚚 markup_text.jsx -> markup_text.tsx

Benjamin-Piper committed:
🚚 Move markup files to markup folder

Benjamin-Piper committed:
♻️ Create markup/index.tsx

mia-entropy commented on web/src/dictionary/markup/to_jsx_child.ts:
```diff
@@ -0,0 +1,111 @@
+    const matched = input.matchAll(regex);
+    for (const match_array of matched) {
+        const whole_string = match_array[0];
+        // Not every capture group of the regex will match. We filter out failed
```
That doesn't sound right. It means in the replacer you could have variables get unexpected matching groups.
I would recommend keeping the matching intact and let the replacer code handle optional parts. This is consistent with most other regex apis.

Also, if we never use whole_string, it may be cleaner to not provide it to the replacer

Benjamin-Piper replied:
addressed

Benjamin-Piper committed:
🧑‍💻 Make make_replacements consistent with regex APIs

Motivation:

- whole_string was unused
- filtering capture groups is inconsistent with regex APIs. By no longer
doing this, we're separating concerns.

Benjamin-Piper committed:
➖ Uninstall htmr

Benjamin-Piper committed:
➕ Install preact-parser

htmr was getting dinged in the typescript compilation of the node
dependencies. Since I didn't want to install react itself, I decided to
move to another library. And preact-parser works directly with preact!

This package is only tested on node version 16. So a warning is produced
when installing it. I personally think that, because this package is
tightly scoped (just parsing), that this isn't worth worrying about.

Benjamin-Piper committed:
➖ Uninstall @types/react

Benjamin-Piper committed:
💡 Shorten comment in replace_string_with_jsx

Benjamin-Piper committed:
🎨 tidy up the code

Benjamin-Piper committed:
🩹 Fix newlines being swallowed by preact-parser
